esphome/components/udp/__init__.py

```diff
@@ -0,0 +1,172 @@
+
+
+def validate_(config):
+    if CONF_ENCRYPTION in config:
+        if CONF_SENSORS not in config and CONF_BINARY_SENSORS not in config:
+            raise cv.Invalid("No sensors or binary sensors to encrypt")
+    elif config[CONF_ROLLING_CODE_ENABLE]:
+        raise cv.Invalid("Rolling code requires an encryption key")
+    if config[CONF_PING_PONG_ENABLE]:
+        if not any(CONF_ENCRYPTION in p for p in config.get(CONF_PROVIDERS) or ()):
+            raise cv.Invalid("Ping-pong requires at least one encrypted provider")
+    return config
+
+
+CONFIG_SCHEMA = cv.All(
+    cv.polling_component_schema("15s")
+    .extend(
+        {
+            cv.GenerateID(): cv.declare_id(UDPComponent),
+            cv.Optional(CONF_PORT, default=18511): cv.port,
+            cv.Optional(CONF_ADDRESSES): cv.ensure_list(cv.ipv4),
+            cv.Optional(CONF_ROLLING_CODE_ENABLE, default=False): cv.boolean,
+            cv.Optional(CONF_PING_PONG_ENABLE, default=False): cv.boolean,
+            cv.Optional(
+                CONF_PING_PONG_RECYCLE_TIME, default="600s"
+            ): cv.positive_time_period_seconds,
+            cv.Optional(CONF_SENSORS): cv.ensure_list(sensor_validation(Sensor)),
+            cv.Optional(CONF_BINARY_SENSORS): cv.ensure_list(
+                sensor_validation(BinarySensor)
+            ),
+            cv.Optional(CONF_PROVIDERS): cv.ensure_list(PROVIDER_SCHEMA),
+        },
+    )
+    .extend(ENCRYPTION_SCHEMA),
+    validate_,
+)
+
+SENSOR_SCHEMA = cv.Schema(
+    {
+        cv.Optional(CONF_REMOTE_ID): cv.string_strict,
+        cv.Required(CONF_PROVIDER): cv.valid_name,
+        cv.GenerateID(CONF_UDP_ID): cv.use_id(UDPComponent),
+    }
+)
+
+
+def require_internal_with_name(config):
+    if CONF_NAME in config and CONF_INTERNAL not in config:
+        raise cv.Invalid("Must provide internal: config when using name:")
+    return config
+
+
+def hash_encryption_key(config: dict):
+    return list(hashlib.sha256(config[CONF_KEY].encode()).digest())
+
+
+async def to_code(config):
+    cg.add_define("USE_UDP")
+    cg.add_global(udp_ns.using)
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    cg.add(var.set_port(config[CONF_PORT]))
+    cg.add(var.set_rolling_code_enable(config[CONF_ROLLING_CODE_ENABLE]))
+    cg.add(var.set_ping_pong_enable(config[CONF_PING_PONG_ENABLE]))
+    cg.add(
+        var.set_ping_pong_recycle_time(
+            config[CONF_PING_PONG_RECYCLE_TIME].total_seconds
+        )
+    )
+    if sensors := config.get(CONF_SENSORS):
+        for sens_conf in sensors:
+            if isinstance(sens_conf, dict):
+                sens_id = sens_conf[CONF_ID]
+                sensor = await cg.get_variable(sens_id)
+                bcst_id = sens_conf.get(CONF_BROADCAST_ID) or sens_id.id
+            else:
+                sensor = await cg.get_variable(sens_conf)
+                bcst_id = sens_conf.id
+            cg.add(var.add_sensor(bcst_id, sensor))
```

Comment:
I am pretty sure `maybe_simple_value` still puts the entire dict in the config object, its simply a validation to allow the user to only specify a value.
Can also simplify the if/for down to one line by providing a default value (`[]`) to `get`.

```suggestion
    for sens_conf in config.get(CONF_SENSORS, []):
        sens_id = sens_conf[CONF_ID]
        sensor = await cg.get_variable(sens_id)
        bcst_id = sens_conf.get(CONF_BROADCAST_ID, sens_id.id)
        cg.add(var.add_sensor(bcst_id, sensor))
```

Reply:
I thought that too, but it failed the first time I tried it. But now it works. I must have screwed something else up.